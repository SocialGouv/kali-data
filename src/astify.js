/*
 * Convert a DILA API structure made of metadata and sections|articles children to a [generic AST tree](https://github.com/syntax-tree/unist#nodes)
 */

// convert to syntax-tree format : flatten articles|sections in children
const astify = (node, depth = 0) => ({
  type: "section",
  data: {
    title: node.title,
    ...(depth === 0 && {
      title: node.titre,
      num: parseInt(node.num, 10),
      shortTitle: (node.categorisation && node.categorisation[0]) || node.title,
      categorisation: node.categorisation
    }),
    // add some data
    intOrdre: node.intOrdre,
    id: node.id,
    etat: node.etat,
    ...(node.modifDate && { modifDate: node.modifDate })
    // add more data when its the root container
  },
  children: [
    ...((node.sections && node.sections.map(node => astify(node, depth + 1))) ||
      []),
    ...((node.articles &&
      node.articles.map(article => ({
        type: "article",
        data: article
      }))) ||
      [])
  ]
});

export default astify;
