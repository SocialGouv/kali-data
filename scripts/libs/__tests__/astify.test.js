import sampleConvention from "./fixtures/sample.json";
import astify, {cleanAst, isValidSection, latestVersionFilter } from "../astify";

test("should convert structure to AST tree", () => {
  expect(astify(sampleConvention)).toMatchSnapshot();
});

test("should clean an AST tree", () => {
  const ast = astify(sampleConvention);
  expect(cleanAst(ast)).toMatchSnapshot();
});

const sampleArticles = [
  {
    cid: 1,
    id: "KALIARTI000001",
  },
  {
    cid: 1,
    id: "KALIARTI000003",
  },
  {
    cid: 1,
    id: "KALIARTI000002",
  },
];

describe("latestVersionFilter", () => {
  test("latestVersionFilter should return false if its an old version of the article", () => {
    expect(
      latestVersionFilter(
        {
          cid: 1,
          id: "KALIARTI000001",
        },
        0,
        sampleArticles,
      ),
    ).toMatchSnapshot();
  });

  test("latestVersionFilter should return true if its the newer version of the article", () => {
    expect(
      latestVersionFilter(
        {
          cid: 1,
          id: "KALIARTI000003",
        },
        0,
        sampleArticles,
      ),
    ).toMatchSnapshot();
  });
});

describe("isValidSection", () => {
  test("should exclude section ABROGE", () => {
    expect(isValidSection({ etat: "ABROGE" })).toEqual(false);
  });
  test("should exclude section PERIME", () => {
    expect(isValidSection({ etat: "PERIME" })).toEqual(false);
  });
  test("should include section VIGUEUR", () => {
    expect(isValidSection({ etat: "VIGUEUR" })).toEqual(true);
  });
  test("should include section VIGUEUR_ETEN", () => {
    expect(isValidSection({ etat: "VIGUEUR_ETEN" })).toEqual(true);
  });
  test("should include section VIGUEUR_NON_ETEN", () => {
    expect(isValidSection({ etat: "VIGUEUR_NON_ETEN" })).toEqual(true);
  });
  test("should include section without etat", () => {
    expect(isValidSection({ jurisState: "VIGUEUR" })).toEqual(true);
  });
});
