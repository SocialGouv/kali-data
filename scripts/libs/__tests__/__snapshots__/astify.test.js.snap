// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`latestArticleVersionFilter latestArticleVersionFilter should return false if its an old version of the article 1`] = `false`;

exports[`latestArticleVersionFilter latestArticleVersionFilter should return true if its the newer version of the article 1`] = `true`;

exports[`should convert structure to AST tree 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "data": Object {
                    "cid": "2111",
                    "id": "2111",
                    "intOrdre": 555,
                  },
                  "type": "article",
                },
                Object {
                  "data": Object {
                    "cid": "2112",
                    "id": "2112",
                    "intOrdre": 55,
                  },
                  "type": "article",
                },
                Object {
                  "data": Object {
                    "cid": "2113",
                    "id": "2113",
                    "intOrdre": 7,
                  },
                  "type": "article",
                },
              ],
              "data": Object {
                "etat": "VIGUEUR",
                "id": "KALISCTA211",
                "intOrdre": 211,
                "title": "title2",
              },
              "type": "section",
            },
            Object {
              "data": Object {
                "cid": "211",
                "id": "211",
              },
              "type": "article",
            },
            Object {
              "data": Object {
                "cid": "212",
                "id": "212",
              },
              "type": "article",
            },
            Object {
              "data": Object {
                "cid": "213",
                "id": "213",
              },
              "type": "article",
            },
          ],
          "data": Object {
            "etat": "VIGUEUR",
            "id": "KALISCTA21",
            "intOrdre": 21,
            "title": "title2-1",
          },
          "type": "section",
        },
      ],
      "data": Object {
        "etat": "VIGUEUR",
        "id": "KALISCTA2",
        "intOrdre": 2222,
        "title": "title2",
      },
      "type": "section",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "data": Object {
                    "cid": "311",
                    "id": "311",
                  },
                  "type": "article",
                },
              ],
              "data": Object {
                "etat": "VIGUEUR",
                "id": "KALISCTA311",
                "intOrdre": 311,
                "title": "title311",
              },
              "type": "section",
            },
          ],
          "data": Object {
            "etat": "VIGUEUR",
            "id": "KALISCTA31",
            "intOrdre": 31,
            "title": "title3",
          },
          "type": "section",
        },
        Object {
          "children": Array [
            Object {
              "data": Object {
                "cid": "321",
                "id": "321",
              },
              "type": "article",
            },
          ],
          "data": Object {
            "etat": "VIGUEUR",
            "id": "KALISCTA32",
            "intOrdre": 32,
            "title": "title32",
          },
          "type": "section",
        },
        Object {
          "data": Object {
            "cid": "31",
            "id": "3",
          },
          "type": "article",
        },
      ],
      "data": Object {
        "etat": "VIGUEUR",
        "id": "KALISCTA3",
        "intOrdre": 1111,
        "title": "title3",
      },
      "type": "section",
    },
  ],
  "data": Object {
    "categorisation": Array [
      "categorie",
    ],
    "etat": "VIGUEUR",
    "id": "1",
    "intOrdre": 1,
    "num": 292,
    "shortTitle": "categorie",
    "title": "title",
  },
  "type": "section",
}
`;
